File: java/Greedy/Job_Sequencing_Problem.java
package java.Greedy;

import java.util.*;

class Job {
  int id;
  int deadline;
  int profit;

  Job(int id, int deadline, int profit) {
    this.id = id;
    this.deadline = deadline;
    this.profit = profit;
  }
}

public class Job_Sequencing_Problem {
  int[] jobScheduling(Job[] arr, int n) {
    Arrays.sort(arr, (a, b) -> b.profit - a.profit);
    int maxDeadline = 0;
    for (Job job : arr) {
      maxDeadline = Math.max(maxDeadline, job.deadline);
    }
    int[] slots = new int[maxDeadline + 1];
    Arrays.fill(slots, -1);
    int count = 0;
    int totalProfit = 0;
    for (Job job : arr) {
      for (int j = job.deadline; j > 0; j--) {
        if (slots[j] == -1) {
          slots[j] = job.id;
          count++;
          totalProfit += job.profit;
          break;
        }
      }
    }
    return new int[] {
        count, totalProfit
    };
  }
}

File: python/Greedy/Job_Sequencing_Problem.py
def jobScheduling(jobs):
  jobs.sort(key=lambda x: x[2], reverse=True)

  max_deadline = max(job[1] for job in jobs)

  slots = [-1] * (max_deadline + 1)

  count = 0 
  profit = 0 

  for job_id, deadline, gain in jobs:
    for j in range(deadline, 0, -1):
      if slots[j] == -1:
        slots[j] = job_id
        count += 1 
        profit += gain 
        break
  
  return [count, profit]
